### PR TITLE
TAJO-1385 Remove locking on RMContext

### PR DESCRIPTION
**Pull request by navis**
Delegated all worker iteration logic to one thread(WorkerResourceAllocationThread), removing synchronization.

### CODE REVIEW

File: tajo-common/src/main/java/org/apache/tajo/util/BasicFuture.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    if (exception != null) {
+      throw new ExecutionException(exception);
+    }
+    if (result != null) {
+      return result;
+    }
+    throw new TimeoutException("Timed-out");
+  }
+
+  public synchronized void done(T result, Exception ex) {
```
Comment (ykrips) [inline tajo-common/src/main/java/org/apache/tajo/util/BasicFuture.java]: Your patch does not contain the code handling exception on BasicFuture, but this class could be used in future and other peaple could be set both result and ex parameter. Setting both two parameters does not dishevel this function flow, but checking validation on these could notify earlier. How about your thought?

Comment (navis) [inline tajo-common/src/main/java/org/apache/tajo/util/BasicFuture.java]: Ok, I'll separate done method into two.

**Comment by navis** [comment]
Addressed comment

**Comment by ykrips** [comment]
I have found an exception on this patch. Could you check this exception?
```
java.util.concurrent.TimeoutException: Timed-out
	at org.apache.tajo.util.BasicFuture.get(BasicFuture.java:70)
	at org.apache.tajo.master.rm.TajoWorkerResourceManager.getClusterResourceSummary(TajoWorkerResourceManager.java:370)
	at org.apache.tajo.master.QueryCoordinatorService$ProtocolServiceHandler.heartbeat(QueryCoordinatorService.java:116)
	at org.apache.tajo.ipc.QueryCoordinatorProtocol$QueryCoordinatorProtocolService$1.heartbeat(QueryCoordinatorProtocol.java:13756)
	at org.apache.tajo.ipc.QueryCoordinatorProtocol$QueryCoordinatorProtocolService.callMethod(QueryCoordinatorProtocol.java:13920)
	at org.apache.tajo.rpc.AsyncRpcServer$ServerHandler.channelRead(AsyncRpcServer.java:124)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:182)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:130)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
	at java.lang.Thread.run(Thread.java:695)
```

**Comment by navis** [comment]
Made a bug merging on other patches, sorry (fixed).

**Comment by jihoonson** [comment]
Hi @navis, sorry for delayed review.
I've almost finished my review.
Only one thing left is testing on a real cluster.
I'll finish soon.

**Comment by jihoonson** [comment]
Sorry for taking long time to review. I have been quite busy in recent days.
This patch looks good to me. I'll commit now.
+1

**Comment by jihoonson** [comment]
I missed closing this PR. Would you close?

**Pull request closed without merging**